src/js/actions/libraries.js

```diff
@@ -607,20 +608,29 @@ define(function (require, exports) {
 
         var textLayers = selectedLayers.filter(function (l) { return l.isTextLayer(); }),
             vectorLayers = selectedLayers.filter(function (l) { return l.isVector(); }),
-            shapeFillOptions = {
+            transactionOpts = {
+                historyStateInfo: {
+                    name: strings.ACTIONS.APPLY_LIBRARY_COLOR,
+                    target: docAdapter.referenceBy.id(currentDocument.id)
+                }
+            };
+
+        var transaction = descriptor.beginTransaction(transactionOpts),
+            actionOpts = {
+                transaction: transaction,
                 coalesce: false,
                 enabled: true,
                 ignoreAlpha: false
-            };
-
-        // FIXME: Setting font color and fill color at the same time will result in two histroy states.
-        //        We should merge the two states when transaction is supported.
-        var setTextColorPromise = textLayers.isEmpty() ? Promise.resolve() :
-                this.transfer(typeActions.setColor, currentDocument, textLayers, color, true, true),
+            },
+            setTextColorPromise = textLayers.isEmpty() ? Promise.resolve() :
+                this.transfer(typeActions.setColor, currentDocument, textLayers, color, actionOpts),
             setShapeFillColorPromise = vectorLayers.isEmpty() ? Promise.resolve() :
-                this.transfer(shapeActions.setFillColor, currentDocument, vectorLayers, color, shapeFillOptions);
+                this.transfer(shapeActions.setFillColor, currentDocument, vectorLayers, color, actionOpts);
 
-        return Promise.join(setTextColorPromise, setShapeFillColorPromise);
+        return Promise.join(setTextColorPromise, setShapeFillColorPromise)
+            .then(function () {
```

Comment:
endTransaction is going to be the one that calls all the batchPlays, so in cases where one of the actions fail, the failure will not be due to the descriptor, but some other reason. We probably don't want to play anything in those cases, and this will provide a nice stopgap.

So I think `.then` is the right way to go.